Added ok if let lint

KitFreddura opened the pull request:
Hey all,

I added a lint that checks for expressions like:

```rust
if let Some(y) = x.parse().ok()
```
And suggests changing to:

```rust
if let Ok(y) = x.parse()
```

This is my first clippy lint, please let me know if I can refactor anything or make it more idiomatic!

Fix #1185.

mcarton commented on clippy_lints/src/ok_if_let.rs:
```diff
@@ -0,0 +1,49 @@
+use rustc::lint::*;
+use rustc::hir::*;
+use utils::{method_chain_args, span_help_and_lint};
+/// **What it does:*** Checks for unnecessary ok() in if let.
```
We usually add a blank line here

mcarton commented on clippy_lints/src/ok_if_let.rs:
```diff
@@ -0,0 +1,49 @@
+use utils::{method_chain_args, span_help_and_lint};
+/// **What it does:*** Checks for unnecessary ok() in if let.
+///
+/// **Why is this bad?** Calling ok() in if let is unnecessary, instead match on Ok()
```
All code in ``` `` ```

mcarton commented on clippy_lints/src/ok_if_let.rs:
```diff
@@ -0,0 +1,49 @@
+declare_lint! {
+    pub OK_IF_LET,
+    Warn,
+    "usage of ok() in if let statements is unnecessary, match on Ok(expr) instead"
```
Same here

mcarton commented on clippy_lints/src/ok_if_let.rs:
```diff
@@ -0,0 +1,49 @@
+            if print::path_to_string(x) == "Some" { //if using ok() on a Some, kick in lint
+                span_help_and_lint(cx, OK_IF_LET, expr.span,
+                "Matching on `Some` with `ok()` is redundant",
+                "Consider matching on `Ok()` instead");
```
If you feel like it you can try to make an actual suggestion here.
(search for `span_suggestion`)

mcarton commented on clippy_lints/src/ok_if_let.rs:
```diff
@@ -0,0 +1,49 @@
+            let ExprMatch(ref op, ref body, ref source) = expr.node, //test if expr is a match
+            let MatchSource::IfLetDesugar { contains_else_clause: _ } = *source, //test if it is an If Let
+            let PatKind::TupleStruct(ref x, _, _)  = body[0].pats[0].node, //get operation
+            let Some(_) = method_chain_args(op, &["ok"]) //test to see if using ok() method
```
What if my non-`Result` type happens to have an `ok` method? You should at least check the type of the expression is `Result` here.

mcarton commented:
BTW, Travis fails because you need to run `./util/update_lints.py`.
Overall this looks like a good first try modulo the missing type-checking and other small stuffs.

mcarton commented on clippy_lints/src/lib.rs:
```diff
@@ -403,6 +405,7 @@ pub fn register_plugins(reg: &mut rustc_plugin::Registry) {
         no_effect::NO_EFFECT,
         no_effect::UNNECESSARY_OPERATION,
         non_expressive_names::MANY_SINGLE_CHAR_NAMES,
+        ok_if_let::OK_IF_LET,
```
I don't really like that name. How about `IF_LET_SOME_RESULT`?

KitFreddura commented:
Ok all that sounds good! I'm working on the updates right now, is there a way to typecheck an Expr? I'm looking for methods to ensure that the expr in question is of type Result.

KitFreddura commented:
Updated as per requests:  
  
Added type checking to ensure that this lint is only occurring for expressions of the form Result<T,E>.ok() and added basic suggestion format.

KitFreddura commented:
I'm a bit confused as to why the Travis build is failing, it looks like this lint is keeping clippy from compiling.

mcarton commented:
Did you run *rustfmt* in 56a07d4? It did a lot of bad things…

mcarton commented on clippy_lints/src/ok_if_let.rs:
```diff
@@ -0,0 +1,53 @@
+        if_let_chain! {[
+            let ExprMatch(ref op, ref body, ref source) = expr.node, //test if expr is a match
+            let MatchSource::IfLetDesugar { .. } = *source, //test if it is an If Let
+            let ExprMethodCall(_, _, ref result_types) = op.clone().unwrap().node, //get expr from expr.ok()
```
We never need to clone nodes. Here you can remove `clone().unwrap().`.

KitFreddura commented:
Whoops, I forgot I had ```rustfmt``` on every save. I fixed the format issues and removed the ```clone().unwrap()```

KitFreddura commented:
Had to refork the repo, creating new pull request elsewhere. Sorry for the confusion.
